---
title: "Untitled"
author: "William Milligan"
date: "November 26, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Data

```{r, echo=FALSE}
out_path = "C:/Users/William Milligan/Desktop/DeepSeqProject/"
name = 'single'

## load CDS datasets
library(readr)
CDS_counts_Ingolia_RPKM <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/CDS_counts_Ingolia_RPKM.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
colnames(CDS_counts_Ingolia_RPKM)[1] = 'transcript'

CDS_MFE_60bp_normalized <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/CDS_MFE_60bp_normalized.txt", " ", escape_double = FALSE, trim_ws = TRUE)
colnames(CDS_MFE_60bp_normalized)[5]='transcript'

CDS_Prop <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/CDS_Prop.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
colnames(CDS_Prop)[1]='transcript'

TableS1C <- read_csv("C:/Users/William Milligan/Desktop/DeepSeqProject/TableS1C.csv")
colnames(TableS1C)[1] ='transcript'
## Merge CDS Info

CDS_info <- merge(CDS_counts_Ingolia_RPKM,CDS_MFE_60bp_normalized,by='transcript')
CDS_info <- merge(CDS_info,CDS_Prop,by='transcript')
CDS_info <- merge(CDS_info,TableS1C,by='transcript')

CDS_average_count_data <- function(input){
  num1 = as.integer(input[4])
  num1 = num1 + as.integer(input[5])
  num1 = num1 + as.integer(input[3])
  return(num1/3)
}

CDS_info$average_count <- apply(CDS_info,FUN = CDS_average_count_data,MARGIN=1)

## Load uORF datasets and fix names

supp3_names <- function(input){
  start = as.integer(input[3])
  length = as.integer(input[7])+1
  end = as.integer(length*3+start)
  
  name = input[1]
  
  our_name = paste(name,start,end,sep='_')
  
  return(our_name)
}

average_count_data <- function(input){
  num1 = as.integer(input[4])
  num1 = num1 + as.integer(input[5])
  num1 = num1 + as.integer(input[6])
  return(num1/3)
}

Supp3 <- read_csv("C:/Users/William Milligan/Desktop/DeepSeqProject/Supp3.csv")
Supp3$uORF <- apply(Supp3,supp3_names,MARGIN=1)

uORF_counts_Ingolia_RPKM <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/uORF_feature_counts_Ingolia.tsv", "\t", escape_double = FALSE, trim_ws = TRUE)
uORF_counts_Ingolia_RPKM$ID <- vapply(strsplit(uORF_counts_Ingolia_RPKM$ID,split=".0",fixed = TRUE), `[`, 1, FUN.VALUE=character(1))
colnames(uORF_counts_Ingolia_RPKM)[2]='uORF'
uORF_counts_Ingolia_RPKM$average_count = apply(uORF_counts_Ingolia_RPKM,MARGIN=1,FUN=average_count_data)

uORF_MFE_60bp_normalized <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/uORFs_MFE_60bp_normalized.txt", " ", escape_double = FALSE, trim_ws = TRUE)
uORF_MFE_60bp_normalized$uORF <- vapply(strsplit(uORF_MFE_60bp_normalized$uORF,split=".0",fixed = TRUE), `[`, 1, FUN.VALUE=character(1))

uORFs_distance_toCDSstart <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/uORFs_distance_toCDSstart.tsv", "\t", escape_double = FALSE, trim_ws = TRUE)
uORFs_distance_toCDSstart$uORF <- vapply(strsplit(uORFs_distance_toCDSstart$uORF_ID,split=".0",fixed = TRUE), `[`, 1, FUN.VALUE=character(1))


uORF_Prop <- read_delim("C:/Users/William Milligan/Desktop/DeepSeqProject/uORF_Prop.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
colnames(uORF_Prop)[3] = 'uORF'
uORF_Prop$uORF <- vapply(strsplit(uORF_Prop$uORF,split=".0",fixed = TRUE), `[`, 1, FUN.VALUE=character(1))
#uORF_info <- cbind(uORF_Prop,uORF_dist_and_MFE)

## Merge uORF Info
uORF_info <- merge(uORFs_distance_toCDSstart,uORF_MFE_60bp_normalized,by='uORF')
uORF_info <- merge(uORF_info,uORF_counts_Ingolia_RPKM,by='uORF')
uORF_info <- merge(uORF_info,Supp3,by='uORF')
```

Oh wait also do GC content

```{r}
library(stringr)

GC_content <- function(input){
  num_gc = str_count(input[2],pattern = '[g,G,c,C]')
  length = as.integer(input[3])
  
  return(num_gc/length)
}

Ingolia_uORFs <- read.table("C:/Users/William Milligan/Desktop/DeepSeqProject/Ingolia_uORFs.fasta", quote="\"", comment.char="", stringsAsFactors=FALSE)
seq = 1:(dim(Ingolia_uORFs)[1]/2)*2
names = 1:(dim(Ingolia_uORFs)[1]/2)*2-1
uORF_Prop = data.frame(uORF = Ingolia_uORFs[names,1], seq = Ingolia_uORFs[seq,1],stringsAsFactors = FALSE)
uORF_Prop$uORF <- vapply(strsplit(uORF_Prop$uORF,split=".0",fixed = TRUE), `[`, 1, FUN.VALUE=character(1))
uORF_Prop$uORF <- vapply(strsplit(uORF_Prop$uORF,split=">",fixed = TRUE), `[`, 2, FUN.VALUE=character(1))
uORF_Prop$length <- sapply(uORF_Prop$seq,FUN=nchar)
uORF_Prop$gc <- apply(uORF_Prop,MARGIN=1,FUN=GC_content)
uORF_Prop <- uORF_Prop[,-2]

uORF_info <- merge(uORF_info,uORF_Prop,by='uORF')

## Merge all info together
colnames(CDS_info) = paste0('CDS_',colnames(CDS_info))
all_info <- merge(uORF_info,CDS_info,by.x='transcript',by.y='CDS_transcript')
colnames(all_info)
dim(all_info)
```


## Extracting Only Useful Info

We have lots of extra columns so lets only keep the variables we like & get marginal uORF GC content

```{r}
keeps_uORF <- c("dist_uORFend_to_CDS","Len_uORF","MFE.norm","Init Codon [nt]","Frame vs CDS","gc")
keeps_CDS <- c("CDS_Length","CDS_MFE.norm","CDS_GC","CDS_log2 TE")
keeps <- c(keeps_uORF,'Codon',keeps_CDS)

potential_outliers <- which(all_info$`CDS_log2 TE` <= -3 | all_info$`CDS_log2 TE` > 3) 
all_info$`CDS_log2 TE`[potential_outliers] <- NA
all_info <- all_info[complete.cases(all_info),]

if (name == 'dom'){
  remove<-c()
  dom <- c()
  for (i in c(1:dim(all_info)[1])){ 
    match = which(all_info[,1] == all_info[i, 1])
    match_Harr = all_info[match,"# Harr Reads"]
    match <- match[order(match_Harr,decreasing = TRUE)[1]]
    if (match==i){
      dom <- c(dom, i)
    }
  }
  dom_info <- all_info[dom,]
  kept_info <- dom_info[,keeps]
} else {
  
  duplicated_uORFS<-all_info$transcript[duplicated(all_info$transcript)]
  kept_info <- all_info[which(!(all_info$transcript %in% all_info$transcript[duplicated(all_info$transcript)])),keeps]
}

kept_info<-kept_info[complete.cases(kept_info),]


colnames(kept_info)

kept_info$Codon <- kept_info$Codon == 'aug'
kept_info$`Frame vs CDS` <- kept_info$`Frame vs CDS`==0

kept_df <- as.data.frame(kept_info)
kept_df <- kept_df[complete.cases(kept_df),]
colnames(kept_df)[dim(kept_df)[2]] <- "TE"
```

# Correlations
```{r}
# get correlation matrix
# put info in a nice order
kept_df2 <- kept_df[,c(keeps_uORF,'Codon',"CDS_Length","CDS_MFE.norm","CDS_GC","TE")]
og_Colnames <- colnames(kept_df2)
nicecolnames <-  c('uORF Distance to CDS','uORF Length','uORF MFE Normalized','Distance to transcript start','Reading Frame (Binary)','uORF GC','Start Codon (Binary)','CDS Length','CDS MFE Normalized','CDS GC Content','TE')
colnames(kept_df2) <- nicecolnames
cormat <- round(cor(kept_df2),2)
library(reshape2)

# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}

library("Hmisc")
sig_corr<-rcorr(as.matrix(kept_df2), type = "pearson")

# Get upper triangle of the correlation matrix
get_sig_tri <- function(cormat){
  sig_corr = cormat$P
  r_corr = cormat$r
  
  r_corr[lower.tri(r_corr)]<- NA
  r_corr[sig_corr > 0.05] <- NA
  return(r_corr)
}

upper_tri <- get_upper_tri(cormat)
melted_cormat <- melt(upper_tri, na.rm = TRUE)

sig_cormat <- round(get_sig_tri(sig_corr),2)
diag(sig_cormat) <- NA

library(ggplot2)
# Melt the correlation matrix
melted_cormat <- melt(upper_tri, na.rm = TRUE)
melted_sig_cormat <- melt(sig_cormat, na.rm = TRUE)
# Create a ggheatmap
ggheatmap <- ggplot(melted_cormat, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
    name="Pearson\nCorrelation") +
  theme_minimal()+ # minimal theme
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1))+
   theme(axis.text.y = element_text(vjust = 1, 
    size = 12, hjust = 1))+
 coord_fixed()
# Print the heatmap
print(ggheatmap)

ggheatmap + 
geom_text(data=melted_sig_cormat,aes(Var2, Var1, label = value), color = "black", size = 5) +
theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.grid.major = element_blank(),
  panel.border = element_blank(),
  panel.background = element_blank(),
  axis.ticks = element_blank(),
  legend.justification = c(1, 0),
  legend.position = c(0.6, 0.7),
  legend.direction = "horizontal")+
  guides(fill = guide_colorbar(barwidth = 7, barheight = 1,
                title.position = "top", title.hjust = 0.5))

figname1 = paste(name,'one_mouse_correlation_heatmap.jpeg',sep='_')
ggsave(paste(out_path,figname1,sep='/'),plot = last_plot(),device="jpeg")
```

OH NO START CODON CORRELATED WITH OTHER STUFFF :(
```{r}
colnames(kept_df2) <- og_Colnames
#get marginal codon
marginal_codon <- residuals(lm(Codon ~.,data = kept_df2[,which(!(colnames(kept_df2)=='TE'))]))
colnames(kept_df2)

cor.test(marginal_codon,kept_df2$TE)

#get marginal gc
marginal_gc <- residuals(lm(gc ~.,data = kept_df2[,which(!(colnames(kept_df2)=='TE'))]))
cor.test(marginal_gc,kept_df2$TE)

#get marginal frame
marginal_frame <- residuals(lm(`Frame vs CDS` ~.,data = kept_df2[,which(!(colnames(kept_df2)=='TE'))]))
cor.test(marginal_frame,kept_df2$TE)

#get marginal dist
marginal_dist <- residuals(lm(`dist_uORFend_to_CDS` ~.,data = kept_df2[,which(!(colnames(kept_df2)=='TE' ))]))
cor.test(marginal_dist,kept_df2$TE)

#get marginal dist
marginal_distnt <- residuals(lm(`Init Codon [nt]` ~.,data = kept_df2[,which(!(colnames(kept_df2)=='TE' ))]))
cor.test(marginal_distnt,kept_df2$TE)

```
## LINEAR MODEL CONSTRUCTION
```{r}
colnames(kept_df2)<- nicecolnames
 
library("leaps")
model_select <- regsubsets(TE~.,kept_df2,nvmax = 11,nbest=8)
summary_model <- summary(model_select)
top <- order(summary_model$bic,decreasing = FALSE)

figname2= paste(name,'mouse_adjr2.jpeg',sep='_')

jpeg(paste(out_path,figname2,sep='/'))
plot(summary_model$bic[top],xlab='Model Index',ylab='BIC',pch=16)
par(new = TRUE)
plot(summary_model$adjr2[top],xlab=NA,ylab=NA,col='red',axes=FALSE,pch=16)
axis(side=4)
mtext(side = 4, line = 2, 'Adjusted R2')
legend('left',pch=16,legend = c('BIC','adj r^2'),col=c('black','red'))
dev.off()

figname3 = paste(name,'mouse_bic_models.jpeg',sep='_')
jpeg(paste(out_path,figname3,sep='/'))
plot(model_select,scale='bic')
dev.off()

colnames(kept_df2) <- og_Colnames
best_model = summary_model$which[top[1],]
names(best_model)[-1] <- og_Colnames[1:10]
best_model = names(best_model[which(best_model)])[-1]
best_model_lm <- lm(TE~.,kept_df2[,c('TE',best_model)])
summary(best_model_lm)

use_for_model = kept_df2[,c('TE',best_model)]
test_index <- sample(1:dim(use_for_model)[1],as.integer(dim(use_for_model)[1]*0.2))
test <- use_for_model[test_index,]
train <- use_for_model[-test_index,]

temp_lm <- lm(TE~.,train)
summary(temp_lm)
predictions <- predict(temp_lm,test)
cor.test(test$TE,predictions)


just_cds <- lm(TE~.,use_for_model[,c('TE','CDS_GC','CDS_Length')])
cds_predict <- predict(just_cds,test)
cor.test(test$TE,cds_predict)


figname4 = paste(name,'mouse_model_performance.jpeg',sep='_')
jpeg(paste(out_path,figname4,sep='/'))
plot(test$TE,predictions,xlab='Log 2 Translational Efficiency',col='red2',pch=16)
points(test$TE,cds_predict,col='black',pch=16)
abline(lm(test$TE ~ cds_predict), col = 'black',lwd=2)
abline(lm(test$TE ~ predictions), col = 'red',lwd=2)
legend('topleft',c('uORF Model','CDS Model'),pch = 16,col=c('red','black'))
dev.off()

just_cds <- lm(TE~.,use_for_model[,c('TE','CDS_GC','CDS_Length')])

```






